src/types/sgrStack.cpp

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "inc/sgrStack.hpp"
+
+using namespace Microsoft::Console::VirtualTerminal::DispatchTypes;
+
+namespace Microsoft::Console::VirtualTerminal
+{
+    SgrStack::SgrStack() noexcept :
+        _nextPushIndex{ 0 },
+        _numSavedAttrs{ 0 }
+    {
+    }
+
+    void SgrStack::Push(const TextAttribute& currentAttributes,
+                        const VTParameters options) noexcept
+    {
+        AttrBitset validParts;
+
+        try
+        {
+            if (options.empty())
+            {
+                // We save all current attributes.
+                validParts.set(static_cast<size_t>(SgrSaveRestoreStackOptions::All));
+            }
+            else
+            {
+                // Each option is encoded as a bit in validParts. All options (that fit) are
+                // encoded; options that aren't supported are ignored when read back (popped).
+                // So if you try to save only unsupported aspects of the current text
+                // attributes, you'll do what is effectively an "empty" push (the subsequent
+                // pop will not change the current attributes), which is the correct behavior.
+
+                for (size_t i = 0; i < options.size(); i++)
+                {
+                    // TRICKY: The VTParameter could be a fake/materialized parameter, in
+                    // which case it has a default value, and the default value of a
+                    // VTParameter depends on what you are assigning to: if you assign /
+                    // convert directly to a size_t, then the default is "1", but for other
+                    // size_t-sized types, the default is "0". We want "default is 0", so we
+                    // convert to a SgrSaveRestoreStackOptions first.
+                    const SgrSaveRestoreStackOptions option = options.at(i);
+                    const size_t optionAsIndex = static_cast<size_t>(option);
+
+                    // Options must be specified singly; not in combination. Values that are
+                    // out of range will be ignored.
+                    if (optionAsIndex < validParts.size())
+                    {
+                        validParts.set(optionAsIndex);
+                    }
+                }
+            }
+        }
+        catch (std::out_of_range&)
+        {
+            // We should not be able to reach here: we pre-check that everything should be
+            // in range.
+            RaiseFailFastException(nullptr, nullptr, 0);

```

Comment:
nit: prefer wil's `FAIL_FAST_HR` or even `FAIL_FAST_CAUGHT_EXCEPTION` (which is a cool one that transforms any exception)

If this is the only exception we can handle, _and_ we're going to failfast for it ... we could just use a `catch(...)` and FAIL_FAST_CAUGHT ... or suppress it and leave?

Reply:
(I know it's an invariant test that's likely required by our static analyzer ;P)

Reply:
I meant to ask about this (and the other `FAIL_FAST` somewhere else in this PR). Am I right in thinking this is going to terminate the app outright? If so, surely that's a bad idea? We're basically saying this condition is so horrendous that the only possibly course of action is to kill anything the user may have been doing at the time. If you had to choose between some colors being incorrect, or losing a document you'd been working on for several hours, I think most people would pick the former.

Reply:
Right--if the failfast were to execute, then yes, it would kill the app. Definitely unpleasant for users. However, I am asserting this code will never be reached. So the failfast *could* just be an assert... but I am confident. So consider this a "blood assert". :D

Reply:
We at one point changed all our living ASSERTs into FAILFASTs, and it was the worst thing we had done to conhost’s reliability to date. They’re the reason Terminal spontaneously exits to this day! They weren’t harming anyone as asserts, and it turns out we just kept chugging along even with inconsistent state.

Maybe “on error resume next” had a place in some world ... 😁